lib/test_queue/runner.rb

```diff
@@ -427,40 +427,53 @@ def distribute_queue
         else
           sock = @server.accept
           cmd = sock.gets.strip
+
+          cmd =~ /^TOKEN=(\w+)/
+          # If we have a slave from a different test run, don't respond, and it will consider the test run done.
+          case $1
+          # No special handling if the token is valid
+          when @run_token
+          when nil
+            STDERR.puts "*** Worker sent no token to master for run #{@run_token}; ignoring."
+            sock.write("WRONG RUN\n")
+            File.open("/tmp/errors.log", "a") {|f| f.puts "WRONG RUN - nil token"}
+            next
+          else
+            STDERR.puts "*** Worker from run #{run_token} connected to master for run #{@run_token}; ignoring."
+            sock.write("WRONG RUN\n")
+            File.open("/tmp/errors.log", "a") {|f| f.puts "WRONG RUN - wrong token"}
+            next
+          end
+
           case cmd
-          when /^POP/
+          when /^TOKEN=\w+ POP/
             # If we have a slave from a different test run, don't respond, and it will consider the test run done.
             if awaiting_suites?
               sock.write(Marshal.dump("WAIT"))
             elsif obj = @queue.shift
               data = Marshal.dump(obj)
               sock.write(data)
             end
-          when /^SLAVE (\d+) ([\w\.-]+) (\w+)(?: (.+))?/
+          when /^TOKEN=\w+ SLAVE (\d+) ([\w\.-]+) (?: (.+))?/
             num = $1.to_i
             slave = $2
-            run_token = $3
-            slave_message = $4
-            if run_token == @run_token
-              # If we have a slave from a different test run, don't respond, and it will consider the test run done.
-              sock.write("OK\n")
-              remote_workers += num
-            else
-              STDERR.puts "*** Worker from run #{run_token} connected to master for run #{@run_token}; ignoring."
-              sock.write("WRONG RUN\n")
-            end
+            slave_message = $3
+
+            sock.write("OK\n")
+            remote_workers += num
+
             message = "*** #{num} workers connected from #{slave} after #{Time.now-@start_time}s"
             message << " " + slave_message if slave_message
             STDERR.puts message
-          when /^WORKER (\d+)/
+          when /^TOKEN=\w+ WORKER (\d+)/
             data = sock.read($1.to_i)
             worker = Marshal.load(data)
             worker_completed(worker)
             remote_workers -= 1
-          when /^NEW SUITE (.+)/
+          when /^TOKEN=\w+ NEW SUITE (.+)/
```

Comment:
This command comes from inside the Runner itself, so I think you need to update the place that sends it to include the token.